server/src/main/java/org/elasticsearch/discovery/PeerFinder.java

```diff
@@ -239,9 +248,16 @@ public Iterable<DiscoveryNode> getFoundPeers() {
         }
     }
 
-    private List<DiscoveryNode> getFoundPeersUnderLock() {
+    private Collection<DiscoveryNode> getFoundPeersUnderLock() {
         assert holdsLock() : "PeerFinder mutex not held";
-        return peersByAddress.values().stream().map(Peer::getDiscoveryNode).filter(Objects::nonNull).distinct().toList();
```

Comment:
This was taking 150s for the test run alone for the `toList`, the version just returning the set is almost invisible in profiling compared to the cost of ``assert holdsLock();`.